ynr/apps/people/merging.py

```diff
@@ -0,0 +1,202 @@
+import json
+
+from django.conf import settings
+from django.contrib.admin.utils import NestedObjects
+from django.db import IntegrityError, connection, transaction
+
+from candidates.models import (
+    LoggedAction,
+    PersonRedirect,
+    UnsafeToDelete,
+    merge_popit_people,
+)
+from candidates.models.versions import get_person_as_version_data
+from candidates.views.version_data import get_change_metadata, get_client_ip
+
+
+class InvalidMergeError(ValueError):
+    """
+    Raised when merging two people would cause invalid data or some other
+    oddness
+    """
+
+    pass
+
+
+class PersonMerger:
+    """
+    Deals with merging two people, ensuring that no data is lost
+    """
+
+    def __init__(self, dest_person, source_person, request=None):
+        self.dest_person = dest_person
+        self.source_person = source_person
+        self.request = request
+
+    def safe_delete(self, model):
+        collector = NestedObjects(using=connection.cursor().db.alias)
+        collector.collect([model])
+        if len(collector.nested()) > 1:
+            related_objects = "\n\t".join(
+                [repr(m) for m in collector.nested()[1:]]
+            )
+            raise UnsafeToDelete(
+                "Can't delete '{}' with related objects: \n {}".format(
+                    model, related_objects
+                )
+            )
+
+        return model.delete()
+
+    def merge_versions_json(self):
+        # Merge the reduced JSON representations:
+        merge_popit_people(
+            get_person_as_version_data(self.dest_person),
+            get_person_as_version_data(self.source_person),
+        )
+
+        # Make sure the secondary person's version history is appended, so it
+        # isn't lost.
+        dest_person_versions = json.loads(self.dest_person.versions)
+        dest_person_versions += json.loads(self.source_person.versions)
+        self.dest_person.versions = json.dumps(dest_person_versions)
+
+    def merge_person_attrs(self):
+        """
+        Merge attributes on Person from source in to dest
+        """
+
+        for field in settings.SIMPLE_POPOLO_FIELDS:
+            source_value = getattr(self.source_person, field.name, None)
+            dest_value = getattr(self.dest_person, field.name, None)
+            # Assume we want to keep the value from dest_person
+            if not dest_value:
+                setattr(self.dest_person, field.name, source_value)
```

Comment:
The comment and code seem to disagree here. You're saying "Assume we want to keep the value from dest_person" in the comment but then the code assigns `dest_person`'s property `source_value`. What is the intent here?

Reply:
I think the code and the commit are consistent:

1. If the dest person has a value for this field, keep it
2. If the dest person does not have a value for that field, try to set that value from the source person. 

I've just pushed 610e89958257bea08f2c64939c0a4c6e1f93080a to validate this, am I missing something?

Reply:
No you're not. I think I was reading that as

```py
if not dest_value:
    # Assume we want to keep the value from dest_person
    setattr(self.dest_person, field.name, source_value)
```

instead of

```py
# Assume we want to keep the value from dest_person
if not dest_value:
    setattr(self.dest_person, field.name, source_value)
```

It does make sense now.

I think that design decision on how we essentially resolve a merge conflict is quite important though. We usually merge new into old (on the basis that the older ID is probably more likely to be in-use already), but that also means we usually discard newer data in favour of older data if we've got the same field filled in on both objects. That's fine, but probably worth a slightly more verbose explanation here (or somewhere?) - I've only just understood it now.

Reply:
Right, I see. I'll write a longer explanation of this.

I wonder if it's worth us enforcing the convention that the lower ID is kept? I could assign `source_person` and `dest_person` depending on the IDs of the people, meaning we *always* keep the lower ID.

`PersonIdentifier` also keeps the created and modified times, so I could get smarter still and always keep the newer ID.

Reply:
> I wonder if it's worth us enforcing the convention that the lower ID is kept?

I guess it depends on the extent to which its a convention and the extent to which its a rule. If enforcing that in the code is going to mean sometimes we want to do the merge the other way and can't, its probably not that helpful. If we _always_ want to merge high into low, then enforcing it is probably useful. I don't really have a handle on the real-world usage of this feature.